src/write/xcoff.rs

```diff
@@ -0,0 +1,497 @@
+            AddressSize::U8 | AddressSize::U16 | AddressSize::U32 => false,
+            AddressSize::U64 => true,
+        };
+
+        let (hdr_size, sechdr_size, rel_size, sym_size) = if is_64 {
+            (
+                mem::size_of::<xcoff::FileHeader64>(),
+                mem::size_of::<xcoff::SectionHeader64>(),
+                mem::size_of::<xcoff::Rel64>(),
+                mem::size_of::<xcoff::Symbol64>(),
+            )
+        } else {
+            (
+                mem::size_of::<xcoff::FileHeader32>(),
+                mem::size_of::<xcoff::SectionHeader32>(),
+                mem::size_of::<xcoff::Rel32>(),
+                mem::size_of::<xcoff::Symbol32>(),
+            )
+        };
+
+        // Calculate offsets and build strtab.
+        let mut offset = 0;
+        let mut strtab = StringTable::default();
+        // We place the shared address 0 immediately after the section header table.
+        let mut address = 0;
+
+        // XCOFF file header.
+        offset += hdr_size;
+        // Section headers.
+        offset += self.sections.len() * sechdr_size;
+
+        // Calculate size of section data.
+        let mut section_offsets = vec![SectionOffsets::default(); self.sections.len()];
+        for (index, section) in self.sections.iter().enumerate() {
+            let len = section.data.len();
+            let sectype = section.kind;
+            // Section address should be 0 for all sections except the .text, .data, and .bss sections.
+            if sectype == SectionKind::Data
+                || sectype == SectionKind::Text
+                || sectype == SectionKind::UninitializedData
+            {
+                section_offsets[index].address = address;
+            } else {
+                section_offsets[index].address = 0;
+            }
+            section_offsets[index].address = address;
+            if len != 0 {
+                // Set the default section alignment as 4.
+                offset = align(offset, 4);
+                section_offsets[index].data_offset = offset;
+                offset += len;
+                address += offset;
+            } else {
+                section_offsets[index].data_offset = 0;
+            }
+        }
+
+        // Calculate size of relocations.
+        for (index, section) in self.sections.iter().enumerate() {
+            let count = section.relocations.len();
+            if count != 0 {
+                section_offsets[index].reloc_offset = offset;
+                offset += count * rel_size;
+            } else {
+                section_offsets[index].reloc_offset = 0;
+            }
+        }
+
+        // Calculate size of symbols.
+        let mut symbol_offsets = vec![SymbolOffsets::default(); self.symbols.len()];
+        let mut symtab_count = 0;
+        for (index, symbol) in self.symbols.iter().enumerate() {
+            symbol_offsets[index].index = symtab_count;
+            if !(!is_64 && symbol.name.len() <= 8) {
```

Comment:
The double negatives make this too hard too understand.
```suggestion
            if is_64 || symbol.name.len() > 8 {
```